drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
 #![no_std]
 #![no_main]
 
+use drv_lpc55_gpio_api::{Direction, Pin, Value};
 use drv_lpc55_spi as spi_core;
 use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_spi_msg::*;
 use lpc55_pac as device;
 use ringbuf::*;
+
+use crc::{Crc, CRC_32_CKSUM};
+use lpc55_romapi::bootrom;
 use userlib::*;
 
 task_slot!(SYSCON, syscon_driver);
 task_slot!(GPIO, gpio_driver);
 
+// See drv/spi-msg/README.md
+// Messages are received from the Service Processor (SP).
+//
+// Messages are framed at the physical layer by the SP asserting the
+// `Chip Select` (negated) or `CSn` signal.
+//
+// Although SPI supports full duplex communications, the messages carried
+// between the SP and the RoT may take significant time for the RoT to process.
+// Therefore, some form of flow control is needed.
+// The RoT has control of the ROT_IRQ line which is used to indicate that the
+// RoT has a message ready for the SP to read.
+//
+// TODO: SP should be able to send a message to resync the RoT communications.
+// To support that semantic, the RoT will always
+// process incoming bytes even if it is transmitting a previously queued
+// message.
+// TODO: properly test that case.
+//
+// Messages for protocol 1 are composed of a four-byte header:
+//   - a protocol version (0x01)
+//   - the lease significant byte of the payload length
+//   - the most significant byte of the payload length
+//   - a message type
```

Comment:
How is message type encoded?

Reply:
This is a bit of redundant documentation. I'll clean this up or refer to the complete version.

Reply:
added a more detailed description.